Content.Shared/StatusEffect/RefCountSystem.cs

```diff
@@ -0,0 +1,147 @@
+using Robust.Shared.Network;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Timing;
+using Robust.Shared.Utility;
+
+namespace Content.Shared.StatusEffect;
+
+/// <summary>
+/// Manages adding and removing ref-counted components through a public API.
+/// </summary>
+public sealed class RefCountSystem : EntitySystem
+{
+    [Dependency] private readonly IComponentFactory _factory = default!;
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] private readonly INetManager _net = default!;
+
+    #region Public API
+
+    /// <summary>
+    /// Add a ref-counted component by incrementing its sources list.
+    /// You MUST call <see cref="Remove"/> afterwards or the component gets "leaked" and is never removed.
+    /// You also MUST NOT remove the component through normal means or the reference count will be wrong.
+    /// </summary>
+    /// <returns>True if the component was just added</returns>
+    public bool Add<T>(Entity<RefCountComponent?> ent) where T: IComponent, new()
+    {
+        if (!_timing.IsFirstTimePredicted)
+            return false;
+
+        ent.Comp ??= EnsureComp<RefCountComponent>(ent);
+        EnsureComp<T>(ent);
+        return Increment((ent, ent.Comp), typeof(T));
+    }
+
+    /// <summary>
+    /// Add a ref-counted from a reflected component type.
+    /// Same rules apply from the generic version.
+    /// </summary>
+    public bool Add(Entity<RefCountComponent?> ent, Type type, bool force = false)
+    {
+        if (!_timing.IsFirstTimePredicted)
+            return false;
+
+        ent.Comp ??= EnsureComp<RefCountComponent>(ent);
+        if (!Increment((ent, ent.Comp), type))
+            return false;
+
+        var comp = (Component) _factory.GetComponent(type);
+        AddComp(ent, comp, force);
```

Comment:
Isn't this cast pointless.

Reply:
i think it was crying about it before

Reply:
GetComponent returns IComponent but AddComp requires Component 4 no raisin